### PR TITLE
Add AMQP queue handler

### PR DESCRIPTION
**Pull request by esdrasjnr**
What does this implement/fix? Explain your changes.
---------------------------------------------------
This patch adds the AMQP interface to read `data.publish` messages and saves the data in DB

Where has this been tested?
---------------------------
**Operating System/Platform:** Ubuntu 18.04.4
**Go Version:** go1.13.8

### CODE REVIEW

File: go.mod726238372.tmp [hunk 1]
```diff
@@ -0,0 +1,17 @@
+module github.com/CESARBR/knot-cloud-storage
```
Comment (netoax) [inline go.mod726238372.tmp]: Should we add this file?

File: go.mod [hunk 2]
```diff
@@ -3,11 +3,32 @@ module github.com/CESARBR/knot-cloud-storage
 go 1.14
 
 require (
+	github.com/ahmetb/govvv v0.3.0 // indirect
+	github.com/cenkalti/backoff/v4 v4.0.0
+	github.com/cespare/reflex v0.2.0 // indirect
```
Comment (netoax) [inline go.mod]: Is it only my impression or a lot more changes are being added to this file? If that's the case, break into another commit.

File: pkg/entities/data.go [hunk 3]
```diff
@@ -9,6 +11,12 @@ type Data struct {
...
 
 type Payload struct {
-	SensorId int `bson:"sensor_id" json:"sensor_id"`
-	Value    int `bson:"value" json:"value"`
+	SensorId int         `bson:"sensor_id" json:"sensor_id"`
+	Value    interface{} `bson:"value" json:"value"`
+}
+
+// DataPublish represents the incoming publish data command
+type DataPublish struct {
```
Comment (netoax) [inline pkg/entities/data.go]: I think this structure is coupled with the specific event received through `amqp` transport and will be used only to unmarshal this event. It's similar to the network `msgs` we defined in the `babeltower`. Please, move it to a similar file.

File: pkg/server/handler.go [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+	"time"
+
+	"github.com/CESARBR/knot-cloud-storage/pkg/entities"
+	"github.com/CESARBR/knot-cloud-storage/pkg/interactor"
+	"github.com/CESARBR/knot-cloud-storage/pkg/logging"
+	"github.com/CESARBR/knot-cloud-storage/pkg/network"
+)
+
+const (
+	queueNameFogIn        = "storage-in"
```
Comment (netoax) [inline pkg/server/handler.go]: What do you think about renaming to "storage-thing-data". The usage of "in" seems redundant to me because the client only bind queues to an exchange in order to receive (incoming) data.

File: pkg/server/handler.go [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+	if err != nil {
+		return fmt.Errorf("message body parsing error: %w", err)
+	}
+
+	timestamp := time.Now()
+
+	for _, data := range msg.Data {
+		err = h.dataInteractor.Save(entities.Data{From: msg.ID, Payload: data, Timestamp: timestamp})
+		if err != nil {
+			return fmt.Errorf("error saving data: %w", err)
```
Comment (netoax) [inline pkg/server/handler.go]: I think it's important logging this error. What do you think?

File: cmd/main.go [hunk 6]
```diff
@@ -16,16 +19,53 @@ func main() {
...
+		case started := <-handlerStartChan:
+			if started {
+				logger.Info("Handler started")
+			} else {
+				quit <- true
+			}
+		case <-quit:
+			handler.Stop()
+			amqp.Stop()
+			server.Stop()
```
Comment (netoax) [inline cmd/main.go:67]: Nice :)

**Review by netoax — CHANGES_REQUESTED** [review]
Great patch, @esdrasjnr!
Just some initial comments.

File: go.mod726238372.tmp (hunk 1 above)
Comment (lcbm) [reply go.mod726238372.tmp]: Nope, generated `.tmp` files shouldn't be committed :stuck_out_tongue:

File: pkg/server/server.go [hunk 7]
```diff
@@ -29,15 +29,22 @@ func NewServer(port int, logger logging.Logger, dataController *controllers.Data
...
 		s.logger.Error(err)
+		started <- false
 	}
 }
 
+// Stop stops the server
+func (s *Server) Stop() {
+	s.logger.Debug("Server stopped")
+}
+
```
Comment (lcbm) [inline pkg/server/server.go]: This doesn't do anything, actually. Is it really necessary to add it? :eyes: 

You can even take a look at the [babeltower](https://github.com/CESARBR/knot-babeltower/blob/master/pkg/server/server.go) server file as reference, but personally I believe we simply shouldn't add this method.

**Review by lcbm — CHANGES_REQUESTED** [review]
Awesome PR :confetti_ball: 

I made a single observation, and +1'd some of @joaoaneto's comments. Apart from that, LGTM :100:

File: go.mod726238372.tmp (hunk 1 above)
Comment (esdrasjnr) [reply go.mod726238372.tmp]: Sorry :smile: Already remove it!

File: pkg/entities/data.go (hunk 3 above)
Comment (esdrasjnr) [reply pkg/entities/data.go]: Done!

File: pkg/server/handler.go (hunk 4 above)
Comment (esdrasjnr) [reply pkg/server/handler.go]: Agree with you! Done!

File: pkg/server/handler.go (hunk 5 above)
Comment (esdrasjnr) [reply pkg/server/handler.go]: I think it too, but I'm logging in `handler.go` file!

File: cmd/main.go (hunk 6 above)
Comment (esdrasjnr) [reply cmd/main.go:67]: Thanks :smiley:

File: go.mod (hunk 2 above)
Comment (esdrasjnr) [reply go.mod]: As we discussed, this changes have been removed!

File: pkg/controllers/data.go [hunk 8]
```diff
@@ -8,85 +8,88 @@ import (
...
-func respondWithJson(w http.ResponseWriter, code int, payload interface{}) {
+func (d *DataController) respondWithJson(w http.ResponseWriter, code int, payload interface{}) {
 	response, _ := json.Marshal(payload)
 	w.Header().Set("Content-Type", "application/json")
 	w.WriteHeader(code)
-	w.Write(response)
+	_, err := w.Write(response)
+	if err != nil {
+		d.logger.Error(err)
+	}
```
Comment (lcbm) [inline pkg/controllers/data.go:43]: Shouldn't the error be returned? :thinking: 
Same comment as [here](https://github.com/CESARBR/knot-cloud-storage/pull/16#discussion_r407674116).

**Review by lcbm — COMMENTED** [review]
Awesome PR :confetti_ball: 
Left a comment, more of a question. Let me know if anything :smile:

File: pkg/controllers/data.go (hunk 8 above)
Comment (netoax) [reply pkg/controllers/data.go:43]: The main `handler` function is called by the `mux` router and it doesn't require an error to be returned, as we can see [here](https://www.gorillatoolkit.org/pkg/mux#Router.HandleFunc).

File: pkg/entities/data.go (hunk 3 above)
Comment (netoax) [reply pkg/entities/data.go]: @esdrasjnr now it's fine! 
I'd just suggest you move the `msg.go` to the `network` directory.

File: pkg/server/server.go (hunk 7 above)
Comment (netoax) [reply pkg/server/server.go]: @lcbm are you sure there isn't an HTTP server running? 👀
I agree with you this method doesn't do anything, but it'd be nice to clean up gracefully when stopping the server, i.e. @esdrasjnr can follow the reference suggested.

File: pkg/entities/data.go (hunk 3 above)
Comment (esdrasjnr) [reply pkg/entities/data.go]: Done!

**Review by lcbm — APPROVED** [review]
LGTM :confetti_ball:

File: pkg/controllers/data.go (hunk 8 above)
Comment (esdrasjnr) [reply pkg/controllers/data.go:43]: Done!

**Review by netoax — APPROVED** [review]
LGTM!
